uswds/uswds
# USWDS - In-page navigation: Add an option to designate the main content element

# Summary

**Added the optional `data-main-content-selector` attribute to the in-page navigation component.** This attribute allows users to designate which element they want the component to pull headers from. If the  attribute is not defined, the component will pull headers from the `<main>` element.

## Breaking change

This is not a breaking change.

## Related issue

Closes #5050

## Related pull requests

[Changelog and documentation PR](https://github.com/uswds/uswds-site/pull/2190)

  > **Note** 
  > This issue was originally worked in https://github.com/uswds/uswds/pull/5124. 
  >
  > The edits in this new PR address only the ability to select a custom content region. We will need to discuss if we want to incorporate the other work included in this original PR, namely the changes to the component unit tests. 

## Preview link
[In-page navigation](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-in-page-nav-parent/?path=/story/components-in-page-navigation--default)

## Problem statement

In its current state, the in-page navigation component is hard-coded to pull headers from the `<main>` element. Users are not able to customize where the component will look for headers. 

## Solution

This PR creates the option to designate which element the in-page nav will pull headers from. It does this by creating the option to include a `data-main-content-selector` attribute on the `usa-in-page-nav` element. This attribute can accept both class and id values. If the `data-main-content-selector` attribute is not defined, the component will pull from the `<main>` element.

Example usage with a class selector: 
> **Note**
> Using a class selector will allow for the component to pull from multiple content regions on the same page

```html
<aside class="usa-in-page-nav" 
    data-main-content-selector=".main-content">
</aside>
<main>
    <h2>This header will not be included in the component link list</h2>
    <section class="main-content">
      <h2>This header will be included in the component link list</h2>
    </section>
</main>
```

Example usage with an id selector: 

```html
<aside class="usa-in-page-nav" 
    data-main-content-selector="#main-content">
</aside>
<main>
    <h2>This header will not be included in the component link list</h2>
    <section id="main-content">
      <h2>This header will be included in the component link list</h2>
    </section>
</main>
```


## Testing and review
- Confirm that the `data-main-content-selector` attribute name is intuitive 
- Confirm that adding a custom value to the `data-main-content-selector` attribute pulls the expected headings
- Let me know if this should have a unit test (I might need to pair up for that)

To test in Storybook:
1. Open the [Test Custom Content Selector](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-in-page-nav-parent/?path=/story/components-in-page-navigation--test-custom-content-selector) story. Confirm that only this link appears in the in-page navigation:

   ![image](https://github.com/uswds/uswds/assets/93996430/5c5b1406-8505-429e-ad1f-60c4d96e3049)
3. Set the `customContentSelector` control to false and reload the page. Confirm that both links are shown in the in-page navigation:

    ![image](https://github.com/uswds/uswds/assets/93996430/cffb3a94-3a90-413c-848b-49c9ed6d9230)


To test in a project:
1. Install this branch
1. Add [in-page navigation component code](https://designsystem.digital.gov/components/in-page-navigation/#in-page-navigation-code) to a test page.
2. Create a custom main content region that you want the component to pull headers from
4. Add the `data-main-content-selector` to the `usa-in-page-nav` element and define it with the id or class from the custom content region
5. Confirm that the in-page navigation only pulls headers form the designated content region

## Credits
Thanks to @aduth for the suggestion in [this comment](https://github.com/uswds/uswds/pull/5124/files#r1152333721) on PR   https://github.com/uswds/uswds/pull/5124
